### PR TITLE
Marsmedia adapter submission

### PR DESCRIPTION
**Pull request by chensMMG**
First master update

- Add new marsmedia adapter
- Add new marsmedia analytic adapter
- Add new marsmedia spec
- Update adapters.json file
- Update analatics.json file

### CODE REVIEW

**Comment by protonate** [comment]
Please add a title and description for this change.

**Comment by dbemiller** [comment]
@chensMMG please clean up the diff before review. Lots of extraneous whitespace changes to the cox & example files.

**Comment by chensMMG** [comment]
@dbemiller Which diff are you talking about?

**Comment by dbemiller** [comment]
This PR. It shows lots of changes to the cox adapters and `pbjs_example_gpt.html`, which are unrelated to your adapter.

https://github.com/prebid/Prebid.js/pull/1307/files

**Comment by chensMMG** [comment]
@dbemiller I have updated all the files and added contact information. Now I can't make a pull request.

**Comment by dbemiller** [comment]
What do you mean? This is a pull request :).

There are still a ton of unwanted changes to `integrationExamples/gpt/pbjs_example_gpt.html` though

**Comment by chensMMG** [comment]
Yeah i see that now :).
What changes?
I have the latest version of that file, can you please refer me to to the changes?
Although, when I try to run the page I get an error loading: 'https://securepubads.g.doubleclick.net/gampad/ads'.
I have a similar but a little cleaner test page and there everything works fine.

**Comment by dbemiller** [comment]
See [this link](https://github.com/prebid/Prebid.js/pull/1307/files#diff-b90efe89db0168b15565264f69a37b7d) (or click the "files changed" link at the top of github).

The diff is showing a ton of whitespace changes to the file. Are you perhaps on Windows, and your IDE is replacing the unix-style `/n` new lines with windows-style `/r/n` ones?

**Comment by dbemiller** [comment]
Also: can you post the similar-but-cleaner test page?

**Comment by dbemiller** [comment]
Also: I ran your unit tests, and your adapter only has ~50% code coverage. Per [the project standards](https://github.com/prebid/Prebid.js/blob/master/CONTRIBUTING.md#contributing-to-prebidjs), you'll need 80% coverage to be allowed to merge.

**Comment by dbemiller** [comment]
@chensMMG Still waiting for the whitespace changes to be removed from `pbjs_example_gpt.html`.

I'll also need some example bidder params which I can use to test your adapter.

**Comment by chensMMG** [comment]
This is the example bidder for test:
`{
    bidder: 'marsmedia',
     sizes: [[320, 50]],
     params: {
           publisherID: "9999",
           floor: 0
      }
}`

About the whitespaces, I have no errors so I don't know what to fix.

**Comment by dbemiller** [comment]
@chensMMG you're (probably unintentionally) replacing [newline characters](https://stackoverflow.com/a/1279802) in `integrationExamples/gpt/pbjs_example_gpt.html` with Windows-style ones.

You'll know it's fixed when Github shows no changes [to this file](https://github.com/prebid/Prebid.js/pull/1307/files#diff-b90efe89db0168b15565264f69a37b7d) in your PR

File: modules/marsmediaBidAdapter.js [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+      'domain': window.location.hostname
+    };
+
+    return JSON.stringify(params);
+  }
+
+  function getid() {
+    function s4() { return Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1); }
+    return s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() + s4() + s4();
+  }
```
Comment (dbemiller) [inline modules/marsmediaBidAdapter.js]: Theres a `generateUUID` function in `src/utils.js` already.

**Comment by chensMMG** [comment]
Thank you for your comments, I'll make the improvements as soon as possible.
What is your estimation on merging my adapter to the base branch?

**Comment by dbemiller** [comment]
Whenever @mkendall07 comes by and gives it a sanity check (as long as he doesn't find any major issues). Probably not more than a few days

**Comment by chensMMG** [comment]
@dbemiller I have removed 'getid' function from the adapter and now i'm using generateUUID from utils.
In the analytics adapter line 47 to 51 have been removed.
I could not find adapters folder under  test/fixtures.

**Comment by dbemiller** [comment]
@chensMMG That's because the `adapters` folder doesn't exist yet :). If you wanted to, feel free to make it.

Not a blocker from my POV for sure though.

File: modules/marsmediaBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
+import Adapter from 'src/adapter';
+import bidfactory from 'src/bidfactory';
+import bidmanager from 'src/bidmanager';
+import { ajax } from 'src/ajax';
+import { STATUS } from 'src/constants';
+import * as utils from 'src/utils';
+import adaptermanager from 'src/adaptermanager';
+
+const MARS_BIDDER_CODE = 'marsmedia';
+const MARS_BIDDER_URL = '//load3-real12.srv-analytics.info:8080/bidder/?bid=3mhdom';
```
Comment (mkendall07) [inline modules/marsmediaBidAdapter.js]: Is this a product url?

Comment (mkendall07) [inline modules/marsmediaBidAdapter.js]: *prod. Just seems strange to have the port number not be 80

Comment (chensMMG) [inline modules/marsmediaBidAdapter.js]: Yes, I just forgot to remove the port.

Comment (mkendall07) [inline modules/marsmediaBidAdapter.js]: ok. Please update it then and we can merge.

**Comment by dbemiller** [comment]
@chensMMG Do you plan on removing the port 8080 from the `MARS_BIDDER_URL`? Or is it good to merge now?

**Comment by chensMMG** [comment]
We are about to change MARS_BIDDER_URL soon, so better to wait with the merge.
Still the new MARS_BIDDER_URL won't have port 80, I hope it's not a problem.

**Comment by dbemiller** [comment]
Not a problem at all. Just let me know when you're ready for it to merge.

**Comment by chensMMG** [comment]
@dbemiller I have changed MARS_BIDDER_URL to the correct url, now it's ok to merge.

**Comment by dbemiller** [comment]
@chensMMG Merged now. This will go out in the 0.27 release.

You'll probably also want to submit a PR to the [docs repo](https://github.com/prebid/prebid.github.io) to get your bidder added to the [bidders params](http://prebid.org/dev-docs/bidders.html) page, which many publishers use.

**Comment by chensMMG** [comment]
Yes I'm working on the PR.
I have another question:

I have a product that use fraud detection functions,  I need to run prebid.js script and fraud detection simultaneously and show the ad only for clients which are not detected as fraud.
Most important that the ad appears only after the fraud detection is finished.
Do you have any solution for this case?

**Comment by dbemiller** [comment]
Open an Issue about it. You'll get more eyes on it that way, and have a better chance of good answers.

**Comment by chensMMG** [comment]
Ok, thank you.

**Comment by vladi-mmg** [comment]
@dbemiller @dluxemburg

Hi Team, nice to meet you.
I'll take over this project from now, 
Can you tell me what is the current status with our adapter and what is needed for me to upload it correctly?
Thanks

**Comment by jsnellbaker** [comment]
Hi @vladi-mmg 

If the status you're asking for is in terms of the main Prebid.js project, currently only the MarsMedia Analytics Adapter exists in the `master` branch (found [here](https://github.com/prebid/Prebid.js/blob/master/modules/marsmediaAnalyticsAdapter.js)).  

If you want to update your Bid Adapter, you'll need to submit a new PR against the `master` branch.  You can either update your existing bid adapter from the `legacy` branch, or create a new one (not sure which would be easier for you at this point).  

There are a few pages that can help you along with this process:
- http://prebid.org/dev-docs/bidder-adaptor.html (gives an overview of submitting a bid adapter, what's required in the setup, examples, etc)
- http://prebid.org/dev-docs/prebid-1.0-API.html (if you're more familiar with Prebid.js in how used to work, this page can provide an overview of the major changes that occurred in the API during the 1.x update)

These other pages can provide some additional frame of reference on using Prebid.js:
- http://prebid.org/dev-docs/adunit-reference.html
- http://prebid.org/dev-docs/publisher-api-reference.html

If you have any questions on the above, please let me know.  I hope this helps.

**Pull request merged**